File: 2022/Day5.hs
#! /usr/bin/env -S runhaskell
import Control.Monad ((<=<), filterM)
import Data.Array
import Data.Char (isAlpha)
import Data.List
import Data.List.Split
import Data.Maybe (isJust, mapMaybe)
import Text.Read (readMaybe)

import Util

type Stack = String
type Command = (Int, Int, Int)
type Stacks = Array Int Stack

data Input = Input Stacks [Command] deriving (Show)
type Output = String

parse :: String -> Maybe Input
parse s = fmap (Input $ listArray (1,length stacks) stacks) $ parseCommands b
  where
    [a, b] = superlines s
    parseCommands = mapM (tuple3 . mapMaybe readMaybe . splitOn " ") . lines
    parseStacks = filter (not . null) . map (filter isAlpha . trim) . transpose . lines
    stacks = parseStacks a

move :: Bool -> Stacks -> Command -> Stacks
move rev stacks (amount, from, to) = stacks // [(from, stackA), (to, stackB)]
 where
  toMove = (if rev then reverse else id) . take amount $ stacks ! from
  stackA = drop amount $ stacks ! from
  stackB = toMove ++ stacks ! to

solve :: Bool -> Input -> Output
solve isPart1 (Input stacks commands) = map head . elems $ foldl (move isPart1) stacks commands

part1 :: Input -> Output
part1 = solve True

part2 :: Input -> Output
part2 = solve False

main :: IO()
main = do
    f <- readFile("5.in")
    let input = parse f
    case input of
      Nothing -> putStrLn "Unable to read input."
      Just i -> do
        putStr "Solution for part 1: "
        putStrLn . show $ part1 i
        putStr "Solution for part 2: "
        putStrLn . show $ part2 i

-- Solution part 1: SHQWSRBDL
-- Solution part 2: CDTQZHBRS

File: 2022/Day5.py
#! /usr/bin/env python
# pylint: disable=unused-wildcard-import
from util import *
from copy import deepcopy

def parse(inp):
    a, b = superlines(inp)

    g = np.array([list(x) for x in a.splitlines()]).transpose()
    s = [[x for x in l if x != ' '] for l in g[1::4, -2::-1]]
    m = [nums(l) for l in b.splitlines()]

    return s, m

def move(s, amount, src, dst, rev=True):
    src -= 1
    dst -= 1
    m = s[src][-amount:]
    if rev:
        m = list(reversed(m))
    s[src] = s[src][:-amount]
    s[dst] += m

def ans(s):
    return ''.join(map(lambda l: l[-1], s))

def p1(inp):
    s, b = inp

    for x, y, z in b:
        move(s, x, y, z, True)
    return ans(s)

def p2(inp):
    s, b = inp

    for x, y, z in b:
        move(s, x, y, z, False)
    return ans(s)

def main():
    inp = parse(data())
    print(f"Solution for part 1:\n{p1(deepcopy(inp))}")
    print(f"Solution for part 2:\n{p2(inp)}")

if __name__ == "__main__":
    main()

# Solution part 1: SHQWSRBDL
# Solution part 2: CDTQZHBRS
